### PR TITLE
Near 22

### PR DESCRIPTION
**Pull request by m3mayllon**
### Summary
-  `/config/config.json`: add new metrics to plaid models and create plaid 100k params
- `requirements.txt`: update libraries
- `/routers/plaid.py`: add data formatting and update validations
- `/support/helper.py`: add helper functions for plaid templating
- `/support/metrics_covalent.py`: replace '()' to '_' under feedback keys
    - `/tests/test_covalent`: legacy changes
- `/support/metrics_plaid.py`: refactor functions to speed up processing
- `support/models.py`: update plaid model
- `support/score.py`: update plaid score
- `unit_tests.py`: comment out plaid tests (temporarily)

### CODE REVIEW

File: support/helper.py [hunk 1]
```diff
@@ -123,15 +142,15 @@ def filter_dict(lst, key, value):
...
+def dict_reverse_cumsum(lst, col, sum_col):
+    df = pd.DataFrame(lst)
     cols = df.columns
     data = pd.DataFrame(columns=cols)
     accounts = df['account_id'].unique().tolist()
     for account_id in accounts:
         temp = df[df['account_id'] == account_id]
         row = pd.DataFrame(temp[-1:].values, columns=cols)
-        row.at[0, col] = row.at[0, sum_col]*-1
+        row.at[0, col] = row.at[0, sum_col]
```
Comment (irene-bbox) [inline support/helper.py:153]: Please, let's document our **sign convention** in our Git README.md. 

- negative sign = inflow = income --> money flowing into a bank account (e.g., payroll)
- positive sign = outflow = expenses --> money flowing out of bank account (e.g., purchase)

Or vice versa.

File: support/helper.py [hunk 2]
```diff
@@ -141,63 +160,216 @@ def dict_reverse_cumsum(d, col, sum_col):
...
+    metadata['credit_card']['util_ratio']['general']['avg_monthly_value'] = temp['util_ratio'].mean()
+    metadata['credit_card']['util_ratio']['general']['month_count'] = len(data['util_ratio'])
     return metadata
 
 
-def balances(metadata, lst, key, df=None):
-    metadata[key]['balances'] = {}
-    metadata[key]['accounts'] = {}
-    current, limit, high_balance = [], [], []
+def general(metadata, lst, k1, k2='general', df=None):
```
Comment (irene-bbox) [inline support/helper.py:185]: Idea: we could **add docstrings** for long functions like this one that have two or more function parameters.

File: support/helper.py [hunk 3]
```diff
@@ -141,63 +160,216 @@ def dict_reverse_cumsum(d, col, sum_col):
...
+    if data:
+        for p in period:
+            metadata['credit_card']['late_payment']['period'][p] = len([d for d in data if d['timespan'] <= p])
+        metadata['credit_card']['late_payment']['general']['total_count'] = len(data)
+        metadata['credit_card']['late_payment']['general']['month_count'] = len(
+            aggregate_dict_by_month(data, {'amount': ['count']}))
+    return metadata
+
+
+def income(metadata, lst, key, value, k1='checking', k2='income'):
```
Comment (irene-bbox) [inline support/helper.py:258]: Same here: we might need docstrings for the functions:
`keep_dict`, `expenses`, `investments`, `cash_flow`, `earnings`.

File: support/metrics_plaid.py [hunk 4]
```diff
@@ -189,329 +35,79 @@ def plaid_kyc(acc, txn):
...
-        score = 0
-        feedback['credit']['error'] = str(e)
-
-    finally:
-        return score, feedback
-
-
-# @evaluate_function
-def credit_interest(acc, txn, feedback, params):
+def plaid_credit_metrics(feedback, params, metadata):
```
Comment (irene-bbox) [inline support/metrics_plaid.py:40]: Same here: docstrings needed for a function like this. Especially, considering that it's an open source projects that other developers will possibly fork and edit.

File: support/metrics_plaid.py [hunk 5]
```diff
@@ -189,329 +35,79 @@ def plaid_kyc(acc, txn):
...
-            ]
-
-            feedback['credit']['avg_count_monthly_txn'] = round(mean, 0)
+        if metadata['credit_card']:
+            # read metadata
+            d = metadata['credit_card']['general']
+            acc_count = d['accounts']['total_count']
+            txn_avg_count = d['transactions']['avg_monthly_count']
+            txn_timespan = d['transactions']['timespan']
+            bal_limit = sum(d['balances']['limit'])  # shouldnt be average?
```
Comment (irene-bbox) [inline support/metrics_plaid.py:53]: - No, it **shouldn't** be an average. This metric is looking at how much money (in total) have available monthly. That quantity is given by the **sum** (not the average) a user's credit limit across all credit cards.

File: support/metrics_plaid.py [hunk 6]
```diff
@@ -952,95 +267,55 @@ def stability_min_running_balance(acc, txn, feedback, params):
...
 
     except Exception as e:
-        score = 0
         feedback['diversity']['error'] = str(e)
 
     finally:
+        num, size = 2, len(score)
+        if size < num:
+            score = fill_list(score, num, size)
+        print(f'\033[36m  -> Diversity:\t{score}\033[0m')
```
Comment (irene-bbox) [inline support/metrics_plaid.py:320]: Time allowing, I'd like to walk through the refactoring changes in `metrics_plaid.py`.

File: unit_tests.py [hunk 7]
```diff
@@ -9,12 +9,12 @@ def suite():
...
-    suite.addTest(unittest.makeSuite(TestMetricStability))
-    suite.addTest(unittest.makeSuite(TestMetricDiversity))
-    suite.addTest(unittest.makeSuite(TestHelperFunctions))
-    suite.addTest(unittest.makeSuite(TestParametrizePlaid))
+    # suite.addTest(unittest.makeSuite(TestMetricCredit))
+    # suite.addTest(unittest.makeSuite(TestMetricVelocity))
+    # suite.addTest(unittest.makeSuite(TestMetricStability))
+    # suite.addTest(unittest.makeSuite(TestMetricDiversity))
+    # suite.addTest(unittest.makeSuite(TestHelperFunctions))
+    # suite.addTest(unittest.makeSuite(TestParametrizePlaid))
```
Comment (irene-bbox) [inline unit_tests.py:17]: [#66] We must unit test again this new codebase.

**Review by irene-bbox — CHANGES_REQUESTED** [review]

File: config/config.json [hunk 8]
```diff
@@ -18,21 +18,21 @@
...
                                     "length": 0.26,
-                                    "livelihood": 0.12
+                                    "livelihood": 0.12,
+                                    "util_ratio": 0.12,
+                                    "interest": 0.05
                                 }
                             },
                             "velocity": {
                                 "weight": 0.2,
                                 "penalty_weight": 0.33,
```
[the remark this answers is not in the document]
Comment (m3mayllon) [reply config/config.json:29]: 100k is currently just a copy

File: support/helper.py (hunk 1 above)
Comment (m3mayllon) [reply support/helper.py:153]: negative means money leaving the account (outflow / expenses), while positive means inflow

File: config/config.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (m3mayllon) [reply config/config.json:32]: @irene-bbox have updated the weights

File: support/metrics_plaid.py (hunk 6 above)
Comment (m3mayllon) [reply support/metrics_plaid.py:345]: @irene-bbox have added 3 new metrics to credit model and config.json

**Pull request merged**